File: clients/javascript/debug-interface.js
const ClientMessage = require('./codegame/client-message');
const DebugState = require('./debugging/debug-state');
const DebugCommand = require('./debugging/debug-command');
const DebugData = require('./debugging/debug-data');

class DebugInterface {
    constructor(stream) {
        this.stream = stream;
    }

    async addPlacedText(position, text, alignment, size, color) {
        await this.add(new DebugData.PlacedText(position, text, alignment, size, color));
    }
    
    async addCircle(position, radius, color) {
        await this.add(new DebugData.Circle(position, radius, color));
    }
    
    async addGradientCircle(position, radius, innerColor, outerColor) {
        await this.add(new DebugData.GradientCircle(position, radius, innerColor, outerColor));
    }
    
    async addRing(position, radius, width, color) {
        await this.add(new DebugData.Ring(position, radius, width, color));
    }
    
    async addPie(position, radius, startAngle, endAngle, color) {
        await this.add(new DebugData.Pie(position, radius, startAngle, endAngle, color));
    }
    
    async addArc(position, radius, width, startAngle, endAngle, color) {
        await this.add(new DebugData.Arc(position, radius, width, startAngle, endAngle, color));
    }
    
    async addRect(bottomLeft, size, color) {
        await this.add(new DebugData.Rect(bottomLeft, size, color));
    }
    
    async addPolygon(vertices, color) {
        await this.add(new DebugData.Polygon(vertices, color));
    }
    
    async addGradientPolygon(vertices) {
        await this.add(new DebugData.GradientPolygon(vertices));
    }
    
    async addSegment(firstEnd, secondEnd, width, color) {
        await this.add(new DebugData.Segment(firstEnd, secondEnd, width, color));
    }
    
    async addGradientSegment(firstEnd, firstColor, secondEnd, secondColor, width) {
        await this.add(new DebugData.GradientSegment(firstEnd, firstColor, secondEnd, secondColor, width));
    }
    
    async addPolyLine(vertices, width, color) {
        await this.add(new DebugData.PolyLine(vertices, width, color));
    }
    
    async addGradientPolyLine(vertices, width) {
        await this.add(new DebugData.GradientPolyLine(vertices, width));
    }

    async add(debugData) {
        await this.send(new DebugCommand.Add(debugData));
    }
    
    async clear() {
        await this.send(new DebugCommand.Clear());
    }
    
    async setAutoFlush(enable) {
        await this.send(new DebugCommand.SetAutoFlush(enable));
    }
    
    async flush() {
        await this.send(new DebugCommand.Flush());
    }

    async send(command) {
        await (new ClientMessage.DebugMessage(command)).writeTo(this.stream);
        await this.stream.flush();
    }

    async getState() {
        await (new ClientMessage.RequestDebugState()).writeTo(this.stream);
        await this.stream.flush();
        return await DebugState.readFrom(this.stream);
    }
}

module.exports = DebugInterface;

File: clients/ruby/debug_interface.rb
class DebugInterface
    def initialize(stream)
        @stream = stream
    end

    def add_placed_text(position, text, alignment, size, color)
        add(Debugging::DebugData::PlacedText.new(position, text, alignment, size, color))
    end
    
    def add_circle(position, radius, color)
        add(Debugging::DebugData::Circle.new(position, radius, color))
    end
    
    def add_gradient_circle(position, radius, inner_color, outer_color)
        add(Debugging::DebugData::GradientCircle.new(position, radius, inner_color, outer_color))
    end
    
    def add_ring(position, radius, width, color)
        add(Debugging::DebugData::Ring.new(position, radius, width, color))
    end
    
    def add_pie(position, radius, start_angle, end_angle, color)
        add(Debugging::DebugData::Pie.new(position, radius, start_angle, end_angle, color))
    end
    
    def add_arc(position, radius, width, start_angle, end_angle, color)
        add(Debugging::DebugData::Arc.new(position, radius, width, start_angle, end_angle, color))
    end
    
    def add_rect(bottom_left, size, color)
        add(Debugging::DebugData::Rect.new(bottom_left, size, color))
    end
    
    def add_polygon(vertices, color)
        add(Debugging::DebugData::Polygon.new(vertices, color))
    end
    
    def add_gradient_polygon(vertices)
        add(Debugging::DebugData::GradientPolygon.new(vertices))
    end
    
    def add_segment(first_end, second_end, width, color)
        add(Debugging::DebugData::Segment.new(first_end, second_end, width, color))
    end
    
    def add_gradient_segment(first_end, first_color, second_end, second_color, width)
        add(Debugging::DebugData::GradientSegment.new(first_end, first_color, second_end, second_color, width))
    end
    
    def add_poly_line(vertices, width, color)
        add(Debugging::DebugData::PolyLine.new(vertices, width, color))
    end
    
    def add_gradient_poly_line(vertices, width)
        add(Debugging::DebugData::GradientPolyLine.new(vertices, width))
    end

    def add(debug_data)
        send(Debugging::DebugCommand::Add.new(debug_data))
    end
    
    def clear()
        send(Debugging::DebugCommand::Clear.new())
    end
    
    def set_auto_flush(enable)
        send(Debugging::DebugCommand::SetAutoFlush.new(enable))
    end
    
    def flush()
        send(Debugging::DebugCommand::Flush.new())
    end

    def send(command)
        Codegame::ClientMessage::DebugMessage.new(command).write_to(@stream)
        @stream.flush()
    end

    def get_state()
        Codegame::ClientMessage::RequestDebugState.new().write_to(@stream)
        @stream.flush()
        return Debugging::DebugState.read_from(@stream)
    end
end